### ISSUE TITLE
[Eager] Error when training LinearRegressor estimator (Tensor.op)

### ISSUE THREAD

**Issue by fdiazgon** [tensorflow/tensorflow#23515]
**System information**
- Have I written custom code: **Yes**.
- OS Platform and Distribution: **Ubuntu 18.04.1 LTS (Bionic Beaver)**.
- TensorFlow installed from: **binary**.
- TensorFlow version: **1.11.0**.
- Python version: **3.6.6**.
- CUDA/cuDNN version: **N/A. CPU only**.
- GPU model and memory: **N/A. CPU only**.
- Other: **Jupyter notebook, 4.4.0**

**Describe the current behavior**

In Eager mode, training a LinearRegressor Estimator calling `model.train(..)` fails with the following error:

```text
Tensor.op is meaningless when eager execution is enabled.
``` 

**Describe the expected behavior**

`model.train(..)` returns without exceptions.

**Code to reproduce the issue**

```python
import tensorflow as tf
# Uncomment the following line after running the script in Graph mode
# tf.enable_eager_execution()

feature_names = ['a', 'b', 'c']
feature_columns = [tf.feature_column.numeric_column(key=name) for name in feature_names]

true_w = [[-2.0], [4.0], [1.0]]
true_b = [0.5]
noise_level = 0.1

batch_size = 32
num_batches = 1000
steps = 100

def synthetic_dataset(w, b, noise_level, batch_size, num_batches):
    return synthetic_dataset_helper(w, b,
                                    tf.shape(w)[0], noise_level, batch_size,
                                    num_batches)


def synthetic_dataset_helper(w, b, num_features, noise_level, batch_size,
                             num_batches):
    def batch(_):
        x = tf.random_normal([batch_size, num_features])
        y = tf.matmul(x, w) + b + noise_level * tf.random_normal([])
        x_dict = {}
        for idx, name in enumerate(feature_names):
            x_dict[name] = x[:, idx]
        return x_dict, y

    return tf.data.Dataset.range(num_batches).map(batch)

dataset = synthetic_dataset(true_w, true_b, noise_level, batch_size, num_batches)

def input_train():
    return (dataset.make_one_shot_iterator().get_next())

model = tf.estimator.LinearRegressor(feature_columns=feature_columns)
model.train(input_fn=input_train, steps=steps)
```

The code is similar to the following official examples: [linear_regression.py](https://github.com/tensorflow/tensorflow/blob/r1.11/tensorflow/examples/get_started/regression/linear_regression.py) and [linear_regression.py (eager)](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/linear_regression/linear_regression.py).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
Is this still an issue for you? I was not able to execute your script by disabling eager execution. Can you please confirm this?

**Comment by fdiazgon** [comment]
Yes. Same error. But you were right in that the previous script did not work after disabling eager execution. I edited the post with a modified version of the script that works in graph mode but not in eager mode.

**Comment by 2wins** [comment]
FYI
The same issue occurs in the following code.
```python
import tensorflow as tf
keras = tf.keras
from keras.layers import Input, Conv2D, MaxPooling2D, Reshape, Dense, add, \
                         GlobalAveragePooling2D, SeparableConv2D, \
                         BatchNormalization, Activation, Flatten
from keras.models import Model
import numpy as np

tf.enable_eager_execution()


def Xception(x):
    x = Conv2D(32, (3, 3),
            strides=(2, 2),
            use_bias=False,
            name='block1_conv1')(x)
    x = BatchNormalization(name='block1_conv1_bn')(x)
    x = Activation('relu', name='block1_conv1_act')(x)
    x = Conv2D(64, (3, 3), use_bias=False, name='block1_conv2')(x)
    x = BatchNormalization(name='block1_conv2_bn')(x)
    x = Activation('relu', name='block1_conv2_act')(x)

    residual = Conv2D(128, (1, 1),
                        strides=(2, 2),
                        padding='same',
                        use_bias=False)(x)
    residual = BatchNormalization()(residual)

    x = SeparableConv2D(128, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block2_sepconv1')(x)
    x = BatchNormalization(name='block2_sepconv1_bn')(x)
    x = Activation('relu', name='block2_sepconv2_act')(x)
    x = SeparableConv2D(128, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block2_sepconv2')(x)
    x = BatchNormalization(name='block2_sepconv2_bn')(x)

    x = MaxPooling2D((3, 3),
                        strides=(2, 2),
                        padding='same',
                        name='block2_pool')(x)
    x = add([x, residual])

    residual = Conv2D(256, (1, 1), strides=(2, 2),
                        padding='same', use_bias=False)(x)
    residual = BatchNormalization()(residual)

    x = Activation('relu', name='block3_sepconv1_act')(x)
    x = SeparableConv2D(256, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block3_sepconv1')(x)
    x = BatchNormalization(name='block3_sepconv1_bn')(x)
    x = Activation('relu', name='block3_sepconv2_act')(x)
    x = SeparableConv2D(256, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block3_sepconv2')(x)
    x = BatchNormalization(name='block3_sepconv2_bn')(x)

    x = MaxPooling2D((3, 3), strides=(2, 2),
                        padding='same',
                        name='block3_pool')(x)
    x = add([x, residual])

    residual = Conv2D(728, (1, 1),
                        strides=(2, 2),
                        padding='same',
                        use_bias=False)(x)
    residual = BatchNormalization()(residual)

    x = Activation('relu', name='block4_sepconv1_act')(x)
    x = SeparableConv2D(728, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block4_sepconv1')(x)
    x = BatchNormalization(name='block4_sepconv1_bn')(x)
    x = Activation('relu', name='block4_sepconv2_act')(x)
    x = SeparableConv2D(728, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block4_sepconv2')(x)
    x = BatchNormalization(name='block4_sepconv2_bn')(x)

    x = MaxPooling2D((3, 3), strides=(2, 2),
                        padding='same',
                        name='block4_pool')(x)
    x = add([x, residual])

    for i in range(8):
        residual = x
        prefix = 'block' + str(i + 5)

        x = Activation('relu', name=prefix + '_sepconv1_act')(x)
        x = SeparableConv2D(728, (3, 3),
                            padding='same',
                            use_bias=False,
                            name=prefix + '_sepconv1')(x)
        x = BatchNormalization(name=prefix + '_sepconv1_bn')(x)
        x = Activation('relu', name=prefix + '_sepconv2_act')(x)
        x = SeparableConv2D(728, (3, 3),
                            padding='same',
                            use_bias=False,
                            name=prefix + '_sepconv2')(x)
        x = BatchNormalization(name=prefix + '_sepconv2_bn')(x)
        x = Activation('relu', name=prefix + '_sepconv3_act')(x)
        x = SeparableConv2D(728, (3, 3),
                            padding='same',
                            use_bias=False,
                            name=prefix + '_sepconv3')(x)
        x = BatchNormalization(name=prefix + '_sepconv3_bn')(x)

        x = add([x, residual])

    residual = Conv2D(1024, (1, 1), strides=(2, 2),
                        padding='same', use_bias=False)(x)
    residual = BatchNormalization()(residual)

    x = Activation('relu', name='block13_sepconv1_act')(x)
    x = SeparableConv2D(728, (3, 3),
                            padding='same',
                            use_bias=False,
                            name='block13_sepconv1')(x)
    x = BatchNormalization(name='block13_sepconv1_bn')(x)
    x = Activation('relu', name='block13_sepconv2_act')(x)
    x = SeparableConv2D(1024, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block13_sepconv2')(x)
    x = BatchNormalization(name='block13_sepconv2_bn')(x)

    x = MaxPooling2D((3, 3),
                        strides=(2, 2),
                        padding='same',
                        name='block13_pool')(x)
    x = add([x, residual])

    x = SeparableConv2D(1536, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block14_sepconv1')(x)
    x = BatchNormalization(name='block14_sepconv1_bn')(x)
    x = Activation('relu', name='block14_sepconv1_act')(x)

    x = SeparableConv2D(2048, (3, 3),
                        padding='same',
                        use_bias=False,
                        name='block14_sepconv2')(x)
    x = BatchNormalization(name='block14_sepconv2_bn')(x)
    x = Activation('relu', name='block14_sepconv2_act')(x)
    x = GlobalAveragePooling2D()(x)
    x = Dense(2)(x)


class Net(Model):
    def __init__(self):
        super().__init__()

    def call(self, inputs):
        return Xception(inputs)

    def predict(self, inputs):
        logits = self.x(inputs)
        return tf.argmax(logits , axis=-1)

model = Net()
X = np.random.randn(1, 225, 225, 3).astype(np.float32) # dummy input
model(tf.convert_to_tensor(X))
```

**Comment by alextp** [comment]
The issue here is that, in Estimator, when creating a Dataset in the input_fn, you need to create the entire Dataset in the input_fn, not just the iterator. So if you replace

```
def input_train():
    return (dataset.make_one_shot_iterator().get_next())
```

with

```
def input_train():
    dataset = synthetic_dataset(true_w, true_b, noise_level, batch_size, num_batches)
    return (dataset.make_one_shot_iterator().get_next())
```
your code just works.
